yams/scrobble.py

```diff
@@ -879,18 +870,34 @@ def mpd_watch_track(client, session, config):
                         if len(failed_scrobbles) < 1:
                             # If we don't have any pending scrobbles, try to scrobble this
                             scrobble_succeeded = scrobble_track(
-                                song, start_time, base_url, api_key, api_secret, session
+                                song,
+                                status,
+                                start_time,
+                                base_url,
+                                api_key,
+                                api_secret,
+                                session,
                             )
                             # If we've failed, add it to the list for future scrobbles (and write it to the disk)
                             if not scrobble_succeeded:
-                                record_failed_scrobble(
-                                    song, start_time, failed_scrobbles, cache_file_path
+                                failed_scrobble = make_scrobble(
+                                    song, status, timestamp=start_time
                                 )
+                                if failed_scrobble not in failed_scrobbles:
+                                    failed_scrobbles.append(failed_scrobble)
+                                    save_failed_scrobbles_to_disk(
+                                        cache_file_path, failed_scrobbles
+                                    )
                         else:
                             # If we have failed and queued up scrobbles, add this one to the list and try to do them all in one go
-                            record_failed_scrobble(
-                                song, start_time, failed_scrobbles, cache_file_path
+                            failed_scrobble = make_scrobble(
+                                song, status, timestamp=start_time
                             )
+                            if failed_scrobble not in failed_scrobbles:
+                                failed_scrobbles.append(failed_scrobble)
+                                save_failed_scrobbles_to_disk(
+                                    cache_file_path, failed_scrobbles
+                                )
```

Comment:
I wonder if we could squash these repeating lines into a `record_failed_scrobble` function? The biggest reason for that function was simply to prevent repeating lines and to slightly improve legibility.  Just a minor niggle.

I'll probably re-create that function after the merge and add lines 883-890/896-900 into it - if that's all right with you? (or if you want to do it and I can wait on the PR)

Reply:
Yeah, you could - that's what was left of the pre-existing `record_failed_scrobble` function before I removed it. The only thing is, I wasn't sure about whether the check for if `failed_scrobble` was in `failed_scrobbles` was needed, as from what I can see, each `failed_scrobble` has a unique timestamp, and so shouldn't match with anything in `failed_scrobbles` anyway (unless the user's been messing about with their clock). If so, both 883-890 and 896-900 could be reduced to:

```python
failed_scrobbles.append(make_scrobble(
    song, status, timestamp=start_time
)
save_failed_scrobbles_to_disk(
    cache_file_path, failed_scrobbles
)
```

Which seemed a little trivial to make an entire function out of. However, if the check is necessary, I'd be fine with you recreating that function, it'd make that section clearer.